### PR TITLE
Migrate messages on separate pallet

### PR DESCRIPTION
**Pull request by breathx**

### CODE REVIEW

**Review by SabaunT — CHANGES_REQUESTED** [review]
Substrate idiomatic changes - cool! :+1:

File: common/src/storage/messenger.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+use super::*;
+
+/// Message processing centralized behaviour.
```
[the remark this answers is not in the document]
Comment (gshep) [reply common/src/storage/messenger.rs:3]: both are correct )

[the remark this answers is not in the document]
Comment (SabaunT) [reply common/src/storage/messenger.rs:3]: oopsey

File: pallets/gear/src/lib.rs [hunk 2]
```diff
@@ -685,7 +637,7 @@ pub mod pallet {
         pub fn process_queue() -> Weight {
             let mut ext_manager = ExtManager::<T>::default();
 
-            let weight = Self::gas_allowance() as Weight;
+            let weight = GasPallet::<T>::gas_allowance() as Weight;
```
Comment (gshep) [inline pallets/gear/src/lib.rs:640]: I would like this small refactoring to be in a separate PR

File: pallets/gear/src/lib.rs [hunk 3]
```diff
@@ -699,8 +651,13 @@ pub mod pallet {
                 T::DebugInfo::remap_id();
             }
 
-            while Self::can_process_queue() {
-                if let Some(dispatch) = common::dequeue_dispatch() {
+            while <MessengerPallet<T> as Messenger>::QueueProcessing::allowed() {
+                if let Some(dispatch) = <MessengerPallet<T> as Messenger>::Queue::pop_front()
+                    .unwrap_or_else(|_| {
+                        // Can be called only in case of storage corruption
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear/src/lib.rs]: > All these `unreachables` make me sad. Can we eliminate them somehow?

They are seems logical for me, because this is the end of some algorithmic invariants, which are unreachable if we don't force some bad mutates to the storage

[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear/src/lib.rs]: > maybe a bit more details about that? You can write a "proof" in one place and refer to it

I would like to migrate anything and provide big readme with explanations how all pallets works some time soon

File: pallets/gear-messenger/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    pub enum Event<T> {}
+
+    #[pallet::error]
+    pub enum Error<T> {
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear-messenger/src/lib.rs:70]: Do you mean to copy-paste DequeError here?
There will be other errors related to mailbox, WL, etc

File: common/src/storage/deck.rs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+#[derive(Decode, Encode, TypeInfo)]
+pub struct Node<K, V> {
+    pub next: Option<K>,
+    pub value: V,
+}
+
+/// Stripped deck implementation based on map-storage.
+pub trait StorageDeck: Sized {
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/deck.rs]: Exactly, `Deque`

File: pallets/gear/src/lib.rs (hunk 2 above)
Comment (breathx) [reply pallets/gear/src/lib.rs:641]: Yeah, this one is really small. Don't want to waste time for it by creating separate PR. This changes seems reasonable and balanced between other changes. How do you feel?

File: common/src/storage/counter.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+/// Substrate StorageValue simplified interface for counting.
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/counter.rs:1]: I really want to do it some time after I finish changes that I planned before.
https://github.com/gear-tech/gear/pull/874#discussion_r863707020
When I will be having the final version of such traits, I would like to provide common docs and use cases.

Are you satisfied with such plan?

File: common/src/storage/deck.rs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+            if Self::Elements::set(new_tail_key, Node { next: None, value }).is_some() {
+                return Err(DeckError::DuplicateElementKey.into());
+            }
+        } else {
+            let first_key = Self::Key::first(&value);
+
+            let _ = Self::TailKey::set(first_key.clone());
+
+            if Self::HeadKey::set(first_key.clone()).is_some() {
+                return Err(DeckError::HeadWasEmptyWhileTailNot.into());
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/deck.rs]: Exactly, `TailWasEmptyWhileHeadNot`

File: common/src/storage/deck.rs [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                Node {
+                    next: Some(head_key),
+                    value,
+                },
+            )
+            .is_some()
+            {
+                return Err(DeckError::DuplicateElementKey.into());
+            }
+        } else if let Some(tail_key) = Self::TailKey::remove() {
```
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/deck.rs]: Yeah. That branch replaced with another one, which will be returning error while head is empty, but the tail not.

File: pallets/gear-messenger/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear-messenger/src/lib.rs:70]: I would also like to add variable logging in these algos for debugging

File: common/src/storage/deck.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply common/src/storage/deck.rs]: but is it reachable?

File: common/src/storage/counter.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply common/src/storage/counter.rs:1]: Sure. Could you open an issue for that? Just not to forget.

File: pallets/gear-messenger/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        }
+    }
+
+    pub struct DeckImpl<T>(PhantomData<T>);
+
+    pub struct OnPopFrontCallback<T>(PhantomData<T>);
+
+    impl<T: Config> GearCallback<<DeckImpl<T> as GearStorageDeck>::Value> for OnPopFrontCallback<T> {
+        fn call(_arg: &<DeckImpl<T> as GearStorageDeck>::Value) {
+            <Pallet<T> as GearMessenger>::Processed::increase();
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear-messenger/src/lib.rs]: Replaced with `Dequeued`

File: pallets/gear-messenger/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply pallets/gear-messenger/src/lib.rs:69]: I think in order to avoid copy-paste of DequeError and to make pallet errors more precise, we can make a generic Error type in `StorageDeque` trait with some trait bound like `DequeError` to be implemented by `pallet_messenger::Error`. What do you think?
Actually if it's an over-featuring, then no need for that. I mean reaching precise pallet errors should not be so costly.

File: pallets/gear-messenger/src/tests.rs [hunk 10]
```diff
@@ -0,0 +1 @@
+
```
Comment (SabaunT) [inline pallets/gear-messenger/src/tests.rs:1]: maybe some simple tests?

**Review by SabaunT — CHANGES_REQUESTED** [review]

File: pallets/gear-messenger/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+            // Removes value from storage. Single DB write.
+            <Self as GearMessenger>::Sent::clear();
+            weight += T::DbWeight::get().writes(1);
+
+            // Removes value from storage. Single DB write.
+            <Self as GearMessenger>::Dequeued::clear();
+            weight += T::DbWeight::get().writes(1);
+
+            // Puts value in storage. Single DB write.
+            <Self as GearMessenger>::QueueProcessing::allow();
```
Comment (SabaunT) [inline pallets/gear-messenger/src/lib.rs:356]: isn't it set to `true` by default? (`ConstBool<true>` is used, right?)

File: common/src/storage/counter.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/counter.rs:1]: #899

File: pallets/gear/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear/src/lib.rs]: #899

File: pallets/gear/src/lib.rs [hunk 12]
```diff
@@ -103,18 +102,18 @@ impl DebugInfo for () {
...
-    use crate::manager::{ExtManager, HandleKind};
-
     #[pallet::config]
     pub trait Config:
         frame_system::Config
         + pallet_authorship::Config
         + pallet_timestamp::Config
         + pallet_gear_program::Config<Currency = <Self as Config>::Currency>
+        + pallet_gas::Config
+        + pallet_gear_messenger::Config
```
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear/src/lib.rs:116]: Discussed.
It seems not such careful as we thought before due to associated types. Will consider such refactoring in future.

File: common/src/storage/deck.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (breathx) [reply common/src/storage/deck.rs]: > but is it reachable? Seems your code is safe against that. Or..?

We replace our traits api for fallible ones due to the chance of an error (by our algorithm or storage corruption on the user side). Tests should cover everything, but extra checks are never superfluous.

File: pallets/gear-messenger/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (breathx) [reply pallets/gear-messenger/src/lib.rs:69]: I think it's over-engineering for that time. We may run into many troubles and more code len with such method. But will consider some time soon.

File: common/src/storage/deck.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (SabaunT) [reply common/src/storage/deck.rs]: could you please state it in the comment for that `else` block?

**Review by SabaunT — APPROVED** [review]
:fire:

**Review by gshep — APPROVED** [review]
Cool! :guitar:
